File: core/action_group.py
import os
import json
import copy
import base64
from dataclasses import dataclass, field
from typing import List, Dict, Any, Optional, Tuple
from .actions import Action


@dataclass
class ActionGroup:
    name: str
    description: str = ""
    actions: List[Action] = field(default_factory=list)
    
    def to_dict(self) -> Dict[str, Any]:
        return {
            'name': self.name,
            'description': self.description,
            'actions': [action.to_dict() for action in self.actions]
        }
    
    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> 'ActionGroup':
        return cls(
            name=data.get('name', '未命名'),
            description=data.get('description', ''),
            actions=[Action.from_dict(a) for a in data.get('actions', [])]
        )
    
    def get_action_count(self) -> int:
        return len(self.actions)
    
    def validate(self) -> Tuple[bool, str]:
        if not self.name or not self.name.strip():
            return False, "动作组名称不能为空"
        
        if len(self.actions) == 0:
            return False, "动作组必须包含至少一个动作"
        
        for i, action in enumerate(self.actions):
            is_valid, error_msg = action.validate()
            if not is_valid:
                return False, f"动作 {i + 1} 验证失败: {error_msg}"
        
        return True, ""
    
    def validate_dependencies(self) -> Tuple[bool, List[str]]:
        missing_deps = []
        for action in self.actions:
            from .actions import ActionType
            if action.action_type == ActionType.ACTION_GROUP_REF:
                group_name = action.params.get('group_name', '')
                if group_name:
                    missing_deps.append(group_name)
        return len(missing_deps) == 0, missing_deps


class LocalActionGroupManager:
    def __init__(self):
        self._groups: Dict[str, ActionGroup] = {}
    
    def get_all_groups(self) -> List[ActionGroup]:
        seen_names = set()
        unique_groups = []
        for group in self._groups.values():
            if group.name not in seen_names:
                seen_names.add(group.name)
                unique_groups.append(group)
        return unique_groups
    
    def get_group(self, name: str) -> Optional[ActionGroup]:
        return self._groups.get(name)
    
    def has_group(self, name: str) -> bool:
        return name in self._groups
    
    def save_group(self, group: ActionGroup) -> bool:
        is_valid, error_msg = group.validate()
        if not is_valid:
            print(f"动作组验证失败: {error_msg}")
            return False
        self._groups[group.name] = group
        return True
    
    def delete_group(self, name: str) -> bool:
        if name not in self._groups:
            return False
        del self._groups[name]
        return True
    
    def create_group_from_actions(self, name: str, description: str, actions: List[Action]) -> ActionGroup:
        group = ActionGroup(
            name=name,
            description=description,
            actions=[copy.deepcopy(a) for a in actions]
        )
        return group
    
    def get_actions_copy(self, name: str) -> List[Action]:
        group = self.get_group(name)
        if group:
            return [copy.deepcopy(a) for a in group.actions]
        return []
    
    def to_dict(self) -> Dict[str, Any]:
        return {name: group.to_dict() for name, group in self._groups.items()}
    
    def load_from_dict(self, data: Dict[str, Any]) -> Tuple[int, int]:
        self._groups.clear()
        success_count = 0
        fail_count = 0
        for name, group_data in data.items():
            try:
                group = ActionGroup.from_dict(group_data)
                is_valid, error_msg = group.validate()
                if is_valid:
                    self._groups[name] = group
                    success_count += 1
                else:
                    print(f"动作组 '{name}' 验证失败: {error_msg}")
                    fail_count += 1
            except Exception as e:
                print(f"加载动作组 '{name}' 失败: {e}")
                fail_count += 1
        return success_count, fail_count
    
    def clear(self):
        self._groups.clear()
    
    def ensure_group_available(self, name: str, global_manager: 'GlobalActionGroupManager' = None) -> Optional[ActionGroup]:
        group = self.get_group(name)
        if group:
            return group
        
        if global_manager:
            global_group = global_manager.get_group(name)
            if global_group:
                self._groups[name] = copy.deepcopy(global_group)
                return self._groups[name]
        
        return None


class GlobalActionGroupManager:
    _instance = None
    
    def __init__(self):
        self._groups: Dict[str, ActionGroup] = {}
        self._groups_dir = os.path.join(os.path.expanduser('~'), '.simpleRPA', 'groups')
        os.makedirs(self._groups_dir, exist_ok=True)
        self._load_all_groups()
    
    @classmethod
    def get_instance(cls) -> 'GlobalActionGroupManager':
        if cls._instance is None:
            cls._instance = cls()
        return cls._instance
    
    @classmethod
    def reset_instance(cls):
        cls._instance = None
    
    def _load_all_groups(self) -> Tuple[int, int]:
        if not os.path.exists(self._groups_dir):
            return 0, 0
        
        success_count = 0
        fail_count = 0
        
        for filename in os.listdir(self._groups_dir):
            if filename.endswith('.json'):
                filepath = os.path.join(self._groups_dir, filename)
                try:
                    with open(filepath, 'r', encoding='utf-8') as f:
                        data = json.load(f)
                    group = ActionGroup.from_dict(data)
                    is_valid, error_msg = group.validate()
                    if is_valid:
                        self._groups[group.name] = group
                        success_count += 1
                    else:
                        print(f"全局动作组 '{group.name}' 验证失败: {error_msg}")
                        fail_count += 1
                except Exception as e:
                    print(f"加载全局动作组失败 {filename}: {e}")
                    fail_count += 1
        
        return success_count, fail_count
    
    def reload_groups(self) -> Tuple[int, int]:
        self._groups.clear()
        return self._load_all_groups()
    
    def get_all_groups(self) -> List[ActionGroup]:
        seen_names = set()
        unique_groups = []
        for group in self._groups.values():
            if group.name not in seen_names:
                seen_names.add(group.name)
                unique_groups.append(group)
        return unique_groups
    
    def get_group(self, name: str) -> Optional[ActionGroup]:
        return self._groups.get(name)
    
    def has_group(self, name: str) -> bool:
        return name in self._groups
    
    def save_group(self, group: ActionGroup) -> bool:
        is_valid, error_msg = group.validate()
        if not is_valid:
            print(f"动作组验证失败: {error_msg}")
            return False
        
        try:
            self._groups[group.name] = group
            
            filepath = os.path.join(self._groups_dir, f"{group.name}.json")
            with open(filepath, 'w', encoding='utf-8') as f:
                json.dump(group.to_dict(), f, ensure_ascii=False, indent=2)
            
            return True
        except Exception as e:
            print(f"保存全局动作组失败: {e}")
            return False
    
    def delete_group(self, name: str) -> bool:
        if name not in self._groups:
            return False
        
        try:
            del self._groups[name]
            
            filepath = os.path.join(self._groups_dir, f"{name}.json")
            if os.path.exists(filepath):
                os.remove(filepath)
            
            return True
        except Exception as e:
            print(f"删除全局动作组失败: {e}")
            return False
    
    def create_group_from_actions(self, name: str, description: str, actions: List[Action]) -> ActionGroup:
        group = ActionGroup(
            name=name,
            description=description,
            actions=[copy.deepcopy(a) for a in actions]
        )
        return group
    
    def get_actions_copy(self, name: str) -> List[Action]:
        group = self.get_group(name)
        if group:
            return [copy.deepcopy(a) for a in group.actions]
        return []
    
    def ensure_group_loaded(self, name: str) -> Optional[ActionGroup]:
        group = self.get_group(name)
        if group:
            return group
        
        filepath = os.path.join(self._groups_dir, f"{name}.json")
        if os.path.exists(filepath):
            try:
                with open(filepath, 'r', encoding='utf-8') as f:
                    data = json.load(f)
                group = ActionGroup.from_dict(data)
                is_valid, error_msg = group.validate()
                if is_valid:
                    self._groups[group.name] = group
                    return group
                else:
                    print(f"动作组 '{name}' 验证失败: {error_msg}")
            except Exception as e:
                print(f"加载动作组 '{name}' 失败: {e}")
        
        return None
    
    def import_group_from_dict(self, data: Dict[str, Any], overwrite: bool = False) -> Tuple[bool, str]:
        try:
            group = ActionGroup.from_dict(data)
            is_valid, error_msg = group.validate()
            if not is_valid:
                return False, f"验证失败: {error_msg}"
            
            if not overwrite and self.has_group(group.name):
                return False, f"动作组 '{group.name}' 已存在"
            
            if self.save_group(group):
                return True, f"动作组 '{group.name}' 导入成功"
            else:
                return False, "保存失败"
        except Exception as e:
            return False, f"导入失败: {e}"


def encode_image_to_base64(image_path: str) -> Optional[str]:
    try:
        with open(image_path, 'rb') as f:
            image_data = f.read()
        return base64.b64encode(image_data).decode('utf-8')
    except Exception as e:
        print(f"编码图片失败: {e}")
        return None


def decode_base64_to_image(base64_data: str, output_path: str) -> bool:
    try:
        image_data = base64.b64decode(base64_data)
        os.makedirs(os.path.dirname(output_path), exist_ok=True)
        with open(output_path, 'wb') as f:
            f.write(image_data)
        return True
    except Exception as e:
        print(f"解码图片失败: {e}")
        return False


def ensure_action_group_available(group_name: str, local_manager: LocalActionGroupManager = None) -> Optional[ActionGroup]:
    global_manager = GlobalActionGroupManager.get_instance()
    
    if local_manager:
        group = local_manager.get_group(group_name)
        if group:
            return group
    
    group = global_manager.get_group(group_name)
    if group:
        if local_manager:
            local_manager._groups[group_name] = copy.deepcopy(group)
        return group
    
    group = global_manager.ensure_group_loaded(group_name)
    if group and local_manager:
        local_manager._groups[group_name] = copy.deepcopy(group)
    
    return group

File: rust-workspace/crates/simpleRPA-core/src/action_group.rs
use crate::actions::Action;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::fs;
use std::path::PathBuf;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ActionGroup {
    pub name: String,
    #[serde(default)]
    pub description: String,
    #[serde(default)]
    pub actions: Vec<Action>,
}

impl ActionGroup {
    pub fn new(name: &str) -> Self {
        Self {
            name: name.to_string(),
            description: String::new(),
            actions: Vec::new(),
        }
    }

    pub fn to_dict(&self) -> serde_json::Value {
        serde_json::json!({
            "name": self.name,
            "description": self.description,
            "actions": self.actions.iter().map(|a| a.to_dict()).collect::<Vec<_>>(),
        })
    }

    pub fn from_dict(data: &serde_json::Value) -> Option<Self> {
        let name = data.get("name")?.as_str()?.to_string();
        let description = data
            .get("description")
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .to_string();
        let actions = data
            .get("actions")
            .and_then(|v| v.as_array())
            .map(|arr| arr.iter().filter_map(|a| Action::from_dict(a)).collect())
            .unwrap_or_default();

        Some(Self {
            name,
            description,
            actions,
        })
    }

    pub fn validate(&self) -> Result<(), String> {
        if self.name.trim().is_empty() {
            return Err("动作组名称不能为空".into());
        }
        if self.actions.is_empty() {
            return Err("动作组必须包含至少一个动作".into());
        }
        for (i, action) in self.actions.iter().enumerate() {
            action
                .validate()
                .map_err(|e| format!("动作 {} 验证失败: {}", i + 1, e))?;
        }
        Ok(())
    }
}

#[derive(Debug, Clone, Default)]
pub struct LocalActionGroupManager {
    groups: HashMap<String, ActionGroup>,
}

impl LocalActionGroupManager {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn get_all_groups(&self) -> Vec<&ActionGroup> {
        self.groups.values().collect()
    }

    pub fn get_group(&self, name: &str) -> Option<&ActionGroup> {
        self.groups.get(name)
    }

    pub fn has_group(&self, name: &str) -> bool {
        self.groups.contains_key(name)
    }

    pub fn save_group(&mut self, group: ActionGroup) -> Result<(), String> {
        group.validate()?;
        self.groups.insert(group.name.clone(), group);
        Ok(())
    }

    pub fn delete_group(&mut self, name: &str) -> bool {
        self.groups.remove(name).is_some()
    }

    pub fn create_group_from_actions(
        &self,
        name: &str,
        description: &str,
        actions: &[Action],
    ) -> ActionGroup {
        ActionGroup {
            name: name.to_string(),
            description: description.to_string(),
            actions: actions.to_vec(),
        }
    }

    pub fn get_actions_copy(&self, name: &str) -> Vec<Action> {
        self.groups
            .get(name)
            .map(|g| g.actions.clone())
            .unwrap_or_default()
    }

    pub fn to_dict(&self) -> serde_json::Value {
        let map: serde_json::Map<String, serde_json::Value> = self
            .groups
            .iter()
            .map(|(name, group)| (name.clone(), group.to_dict()))
            .collect();
        serde_json::Value::Object(map)
    }

    pub fn load_from_dict(&mut self, data: &serde_json::Value) -> (i32, i32) {
        self.groups.clear();
        let mut success_count = 0;
        let mut fail_count = 0;

        if let Some(obj) = data.as_object() {
            for (name, group_data) in obj {
                if let Some(group) = ActionGroup::from_dict(group_data) {
                    if group.validate().is_ok() {
                        self.groups.insert(name.clone(), group);
                        success_count += 1;
                    } else {
                        fail_count += 1;
                    }
                } else {
                    fail_count += 1;
                }
            }
        }

        (success_count, fail_count)
    }

    pub fn clear(&mut self) {
        self.groups.clear();
    }

    pub fn ensure_group_available(
        &mut self,
        name: &str,
        global_manager: &GlobalActionGroupManager,
    ) -> Option<ActionGroup> {
        if let Some(group) = self.groups.get(name) {
            return Some(group.clone());
        }

        if let Some(group) = global_manager.get_group(name) {
            self.groups.insert(name.to_string(), group.clone());
            return Some(group.clone());
        }

        None
    }
}

pub struct GlobalActionGroupManager {
    groups: HashMap<String, ActionGroup>,
    groups_dir: PathBuf,
}

impl GlobalActionGroupManager {
    pub fn new() -> Self {
        let groups_dir = dirs::home_dir()
            .unwrap_or_else(|| PathBuf::from("."))
            .join(".simpleRPA")
            .join("groups");
        let _ = fs::create_dir_all(&groups_dir);

        let mut mgr = Self {
            groups: HashMap::new(),
            groups_dir,
        };
        mgr.load_all_groups();
        mgr
    }

    fn load_all_groups(&mut self) {
        if !self.groups_dir.exists() {
            return;
        }

        if let Ok(entries) = fs::read_dir(&self.groups_dir) {
            for entry in entries.flatten() {
                let path = entry.path();
                if path.extension().and_then(|e| e.to_str()) == Some("json") {
                    if let Ok(content) = fs::read_to_string(&path) {
                        if let Ok(data) = serde_json::from_str::<serde_json::Value>(&content) {
                            if let Some(group) = ActionGroup::from_dict(&data) {
                                if group.validate().is_ok() {
                                    self.groups.insert(group.name.clone(), group);
                                }
                            }
                        }
                    }
                }
            }
        }
    }

    pub fn reload_groups(&mut self) {
        self.groups.clear();
        self.load_all_groups();
    }

    pub fn get_all_groups(&self) -> Vec<&ActionGroup> {
        self.groups.values().collect()
    }

    pub fn get_group(&self, name: &str) -> Option<&ActionGroup> {
        self.groups.get(name)
    }

    pub fn has_group(&self, name: &str) -> bool {
        self.groups.contains_key(name)
    }

    pub fn save_group(&mut self, group: &ActionGroup) -> Result<(), String> {
        group.validate()?;

        let filepath = self.groups_dir.join(format!("{}.json", group.name));
        let json = serde_json::to_string_pretty(&group.to_dict())
            .map_err(|e| format!("JSON序列化失败: {}", e))?;
        fs::write(&filepath, json).map_err(|e| format!("写入文件失败: {}", e))?;

        self.groups.insert(group.name.clone(), group.clone());
        Ok(())
    }

    pub fn delete_group(&mut self, name: &str) -> bool {
        if self.groups.remove(name).is_some() {
            let filepath = self.groups_dir.join(format!("{}.json", name));
            let _ = fs::remove_file(&filepath);
            true
        } else {
            false
        }
    }

    pub fn ensure_group_loaded(&mut self, name: &str) -> Option<ActionGroup> {
        if let Some(group) = self.groups.get(name) {
            return Some(group.clone());
        }

        let filepath = self.groups_dir.join(format!("{}.json", name));
        if filepath.exists() {
            if let Ok(content) = fs::read_to_string(&filepath) {
                if let Ok(data) = serde_json::from_str::<serde_json::Value>(&content) {
                    if let Some(group) = ActionGroup::from_dict(&data) {
                        if group.validate().is_ok() {
                            self.groups.insert(group.name.clone(), group.clone());
                            return Some(group.clone());
                        }
                    }
                }
            }
        }

        None
    }
}

pub fn ensure_action_group_available(
    group_name: &str,
    local_manager: Option<&mut LocalActionGroupManager>,
) -> Option<ActionGroup> {
    let global_manager = GlobalActionGroupManager::new();

    if let Some(ref mgr) = local_manager {
        if let Some(group) = mgr.get_group(group_name) {
            return Some(group.clone());
        }
    }

    if let Some(group) = global_manager.get_group(group_name) {
        if let Some(mgr) = local_manager {
            mgr.save_group(group.clone()).ok();
        }
        return Some(group.clone());
    }

    None
}

pub fn encode_image_to_base64(image_path: &str) -> Option<String> {
    let data = fs::read(image_path).ok()?;
    use base64::Engine;
    Some(base64::engine::general_purpose::STANDARD.encode(&data))
}

pub fn decode_base64_to_image(base64_data: &str, output_path: &str) -> Result<(), String> {
    use base64::Engine;
    let data = base64::engine::general_purpose::STANDARD
        .decode(base64_data)
        .map_err(|e| format!("Base64解码失败: {}", e))?;
    if let Some(parent) = std::path::Path::new(output_path).parent() {
        fs::create_dir_all(parent).map_err(|e| format!("创建目录失败: {}", e))?;
    }
    fs::write(output_path, data).map_err(|e| format!("写入文件失败: {}", e))
}
